posthog/temporal/data_modeling/workflows/execute_dag.py

```diff
@@ -0,0 +1,337 @@
+        dag_id: The DAG that was orchestrated.
+        scheduled_nodes: Total number of nodes that were scheduled for materialization.
+        successful_nodes: Number of nodes that materialized successfully.
+        failed_nodes: Number of nodes that failed to materialize.
+        skipped_nodes: Number of nodes skipped due to upstream failures.
+        duration_seconds: Total duration of the workflow in seconds.
+        node_results: Individual results for each node.
+    """
+
+    dag_id: str
+    scheduled_nodes: int
+    successful_nodes: int
+    failed_nodes: int
+    skipped_nodes: int
+    duration_seconds: float
+    node_results: list[NodeResult]
+
+
+def _get_edge_lookup(edges: list[tuple[str, str]]):
+    edge_lookup = defaultdict(set)
+    for source, target in edges:
+        edge_lookup[target].add(source)
+    return edge_lookup
+
+
+def _get_dependent_lookup(edge_lookup: dict):
+    # builds the inverse of the edge_lookup in the DAG object
+    dependents = defaultdict(set)
+    for target, sources in edge_lookup.items():
+        for source in sources:
+            dependents[source].add(target)
+    return dependents
+
+
+def _get_downstream_lookup(edge_lookup: dict):
+    downstreams = _get_dependent_lookup(edge_lookup)
+
+    def _get_all_downstream(node_id: str, visited: set[str]) -> set[str]:
+        if node_id in visited:
+            return set()
+        visited.add(node_id)
+        result = set(downstreams[node_id])
+        for downstream in list(downstreams[node_id]):
+            result.update(_get_all_downstream(downstream, visited))
+        return result
+
+    for node in list(downstreams.keys()):
+        visited: set[str] = set()
+        downstreams[node] = _get_all_downstream(node, visited)
+
+    return downstreams
+
+
+def _dag_execution_levels(
+    team_id: int,
+    dag_id: str,
+    nodes: list[str],
+    edge_lookup: dict,
+) -> list[list[str]]:
+    """Compute execution levels using kahn's topological sort."""
+    # Initialize in_degree for all nodes, defaulting to 0 for nodes with no dependencies
+    in_degree = {node_id: len(edge_lookup.get(node_id, set())) for node_id in nodes}
+    # inverse of the edge_lookup
+    dependents = _get_dependent_lookup(edge_lookup)
+    levels: list[list[str]] = []
+    remaining = nodes.copy()
+    while remaining:
+        current_level = [node_id for node_id in remaining if in_degree[node_id] == 0]
+        if not current_level:
+            # the only cases where this is possible are an empty DAG or a cycle in the DAG
+            raise EmptyDAGOrCycleError(f"DAG is either empty or contains a cycle: team={team_id} dag={dag_id}")
+        levels.append(current_level)
+        for node_id in current_level:
+            remaining.remove(node_id)
+            for dependent in dependents[node_id]:
+                if dependent in remaining:
+                    in_degree[dependent] -= 1
+    return levels
+
+
+@temporalio.workflow.defn(name="execute-dag")
```

Comment:
nit: a more descriptive name would be nice here - "data-modeling-execute-dag" - just makes it easier to identify workflows in the temporal UI